From 085bf8b4c55c691ed752e931ff94f6e0080b8a98 Mon Sep 17 00:00:00 2001
From: Fred Underwood <underwoodfred@proton.me>
Date: Sun, 22 Dec 2024 07:06:30 +1000
Subject: [PATCH] add second factor to fingerprint unlock

Ported from GrapheneOS 9ac8b683463c; KeyguardViewMediator.onBiometricAuthenticated hunk re-applied by hand (this tree has an extra selected-user check).
---
 .../KeyguardIndicationControllerTest.java     |  11 +-
 .../phone/BiometricsUnlockControllerTest.java |  75 +++++++-
 .../keyguard/KeyguardUpdateMonitor.java       | 172 ++++++++++++++++--
 .../KeyguardUpdateMonitorCallback.java        |   8 +-
 .../biometrics/AuthRippleController.kt        |  21 ++-
 .../ui/binder/KeyguardBouncerViewBinder.kt    |  21 ++-
 .../classifier/FalsingCollectorImpl.java      |   3 +-
 .../keyguard/KeyguardViewMediator.java        |   5 +-
 .../DeviceEntryFingerprintAuthRepository.kt   |   4 +
 .../KeyguardIndicationController.java         |   5 +-
 .../phone/BiometricUnlockController.java      |  70 +++++--
 .../phone/KeyguardBypassController.kt         |  15 +-
 .../KeyguardStatusBarViewController.java      |  12 +-
 .../phone/StatusBarKeyguardViewManager.java   |  31 +++-
 .../policy/KeyguardStateControllerImpl.java   |   7 +-
 .../keyguard/KeyguardUpdateMonitorTest.java   |  29 ++-
 16 files changed, 415 insertions(+), 74 deletions(-)

diff --git a/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/KeyguardIndicationControllerTest.java b/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/KeyguardIndicationControllerTest.java
index cba5a3fa4..0ae165941 100644
--- a/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/KeyguardIndicationControllerTest.java
+++ b/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/KeyguardIndicationControllerTest.java
@@ -73,6 +73,7 @@ import androidx.test.ext.junit.runners.AndroidJUnit4;
 import androidx.test.filters.FlakyTest;
 import androidx.test.filters.SmallTest;
 
+import com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus;
 import com.android.keyguard.TrustGrantFlags;
 import com.android.settingslib.fuelgauge.BatteryStatus;
 import com.android.systemui.dock.DockManager;
@@ -1075,7 +1076,7 @@ public class KeyguardIndicationControllerTest extends KeyguardIndicationControll
         // WHEN face auth succeeds
         when(mKeyguardUpdateMonitor.getIsFaceAuthenticated()).thenReturn(true);
         mController.getKeyguardCallback().onBiometricAuthenticated(0,
-                BiometricSourceType.FACE, false);
+                BiometricSourceType.FACE, false, SecondFactorStatus.Disabled);
 
         // THEN 'face unlocked' then 'press unlock icon to open' message show
         String unlockedByFace = mContext.getString(R.string.keyguard_face_successful_unlock);
@@ -1099,7 +1100,7 @@ public class KeyguardIndicationControllerTest extends KeyguardIndicationControll
         // WHEN face authenticated
         when(mKeyguardUpdateMonitor.getIsFaceAuthenticated()).thenReturn(true);
         mController.getKeyguardCallback().onBiometricAuthenticated(0,
-                BiometricSourceType.FACE, false);
+                BiometricSourceType.FACE, false, SecondFactorStatus.Disabled);
 
         // THEN show 'face unlocked' and 'swipe up to open' messages
         String unlockedByFace = mContext.getString(R.string.keyguard_face_successful_unlock);
@@ -1122,7 +1123,7 @@ public class KeyguardIndicationControllerTest extends KeyguardIndicationControll
         // WHEN face auth is successful
         when(mKeyguardUpdateMonitor.getIsFaceAuthenticated()).thenReturn(true);
         mController.getKeyguardCallback().onBiometricAuthenticated(0,
-                BiometricSourceType.FACE, false);
+                BiometricSourceType.FACE, false, SecondFactorStatus.Disabled);
 
         // THEN show 'face unlocked' and 'swipe up to open' messages
         String unlockedByFace = mContext.getString(R.string.keyguard_face_successful_unlock);
@@ -1144,7 +1145,7 @@ public class KeyguardIndicationControllerTest extends KeyguardIndicationControll
         // WHEN face auth is successful
         when(mKeyguardUpdateMonitor.getIsFaceAuthenticated()).thenReturn(true);
         mController.getKeyguardCallback().onBiometricAuthenticated(0,
-                BiometricSourceType.FACE, false);
+                BiometricSourceType.FACE, false, SecondFactorStatus.Disabled);
 
         // THEN show 'face unlocked' and 'swipe up to open' messages
         String unlockedByFace = mContext.getString(R.string.keyguard_face_successful_unlock);
@@ -1527,7 +1528,7 @@ public class KeyguardIndicationControllerTest extends KeyguardIndicationControll
         when(mKeyguardUpdateMonitor.getUserCanSkipBouncer(getCurrentUser()))
                 .thenReturn(true);
         mController.getKeyguardCallback().onBiometricAuthenticated(0,
-                BiometricSourceType.FACE, false);
+                BiometricSourceType.FACE, false, SecondFactorStatus.Disabled);
         verifyIndicationMessage(
                 INDICATION_TYPE_BIOMETRIC_MESSAGE,
                 mContext.getString(R.string.keyguard_face_successful_unlock));
diff --git a/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/phone/BiometricsUnlockControllerTest.java b/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/phone/BiometricsUnlockControllerTest.java
index a09d14d04..948ee70a9 100644
--- a/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/phone/BiometricsUnlockControllerTest.java
+++ b/packages/SystemUI/multivalentTests/src/com/android/systemui/statusbar/phone/BiometricsUnlockControllerTest.java
@@ -16,6 +16,8 @@
 
 package com.android.systemui.statusbar.phone;
 
+import static com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus.Disabled;
+import static com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus.Enabled;
 import static com.android.systemui.statusbar.phone.BiometricUnlockController.MODE_NONE;
 import static com.android.systemui.statusbar.phone.BiometricUnlockController.MODE_WAKE_AND_UNLOCK;
 
@@ -47,6 +49,7 @@ import androidx.test.filters.SmallTest;
 import com.android.internal.logging.MetricsLogger;
 import com.android.internal.util.LatencyTracker;
 import com.android.keyguard.KeyguardUpdateMonitor;
+import com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus;
 import com.android.keyguard.logging.BiometricUnlockLogger;
 import com.android.systemui.SysuiTestCase;
 import com.android.systemui.biometrics.AuthController;
@@ -145,8 +148,8 @@ public class BiometricsUnlockControllerTest extends SysuiTestCase {
         when(mUpdateMonitor.isDeviceInteractive()).thenReturn(true);
         when(mKeyguardStateController.isFaceEnrolledAndEnabled()).thenReturn(true);
         when(mKeyguardStateController.isUnlocked()).thenReturn(false);
-        when(mKeyguardBypassController.onBiometricAuthenticated(any(), anyBoolean()))
-                .thenReturn(true);
+        when(mKeyguardBypassController.onBiometricAuthenticated(any(), anyBoolean(),
+                any(SecondFactorStatus.class))).thenReturn(true);
         when(mAuthController.isUdfpsFingerDown()).thenReturn(false);
         when(mVibratorHelper.hasVibrator()).thenReturn(true);
         mDependency.injectTestDependency(NotificationMediaManager.class, mMediaManager);
@@ -179,6 +182,53 @@ public class BiometricsUnlockControllerTest extends SysuiTestCase {
         return biometricUnlockController;
     }
 
+    @Test
+    public void onBiometricAuthenticated_fpWithSecondFactorEnabledNotInteractive_showsBouncer() {
+        when(mUpdateMonitor.isDeviceInteractive()).thenReturn(false);
+
+        // Keyguard should always be showing if second factor is enabled.
+        when(mKeyguardStateController.isShowing()).thenReturn(true);
+
+        // These values don't matter when second factor enabled.
+        when(mUpdateMonitor.isDreaming()).thenReturn(false);
+        when(mDozeScrimController.isPulsing()).thenReturn(true);
+        when(mKeyguardStateController.isMethodSecure()).thenReturn(true);
+
+        // Method under test.
+        mBiometricUnlockController.onBiometricAuthenticated(UserHandle.USER_CURRENT,
+                BiometricSourceType.FINGERPRINT, true, Enabled);
+
+        verify(mStatusBarKeyguardViewManager).showPrimaryBouncer(anyBoolean());
+        verify(mStatusBarKeyguardViewManager, never()).notifyKeyguardAuthenticated(anyBoolean());
+        assertThat(mBiometricUnlockController.getMode())
+                .isEqualTo(BiometricUnlockController.MODE_SHOW_BOUNCER);
+    }
+
+    @Test
+    public void onBiometricAuthenticated_fpWithSecondFactorEnabledInteractive_showsBouncer() {
+        when(mUpdateMonitor.isDeviceInteractive()).thenReturn(true);
+
+        // Keyguard should always be showing if second factor is enabled.
+        when(mKeyguardStateController.isShowing()).thenReturn(true);
+
+        // These values don't matter when second factor enabled.
+        when(mUpdateMonitor.isDreaming()).thenReturn(false);
+        when(mDozeScrimController.isPulsing()).thenReturn(true);
+        when(mKeyguardStateController.isMethodSecure()).thenReturn(true);
+        when(mStatusBarKeyguardViewManager.primaryBouncerIsOrWillBeShowing()).thenReturn(false);
+
+        when(mStatusBarKeyguardViewManager.isBouncerShowing()).thenReturn(false);
+
+        // Method under test.
+        mBiometricUnlockController.onBiometricAuthenticated(UserHandle.USER_CURRENT,
+                BiometricSourceType.FINGERPRINT, true, Enabled);
+
+        verify(mStatusBarKeyguardViewManager).showPrimaryBouncer(anyBoolean());
+        verify(mStatusBarKeyguardViewManager, never()).notifyKeyguardAuthenticated(anyBoolean());
+        assertThat(mBiometricUnlockController.getMode())
+                .isEqualTo(BiometricUnlockController.MODE_SHOW_BOUNCER);
+    }
+
     @Test
     public void onBiometricAuthenticated_fingerprintAndBiometricsDisallowed_showPrimaryBouncer() {
         when(mUpdateMonitor.isUnlockingWithBiometricAllowedSafe(true /* isStrongBiometric */))
@@ -402,7 +452,8 @@ public class BiometricsUnlockControllerTest extends SysuiTestCase {
         reset(mKeyguardBypassController);
         when(mUpdateMonitor.isUnlockingWithBiometricAllowedSafe(anyBoolean())).thenReturn(true);
         when(mKeyguardBypassController.getBypassEnabled()).thenReturn(true);
-        when(mKeyguardBypassController.onBiometricAuthenticated(any(), anyBoolean()))
+        when(mKeyguardBypassController.onBiometricAuthenticated(any(), anyBoolean(),
+                any(SecondFactorStatus.class)))
                 .thenReturn(true);
         when(mStatusBarKeyguardViewManager.primaryBouncerIsOrWillBeShowing()).thenReturn(true);
         // the value of isStrongBiometric doesn't matter here since we only care about the returned
@@ -466,6 +517,21 @@ public class BiometricsUnlockControllerTest extends SysuiTestCase {
         captor.getValue().run();
     }
 
+    @Test
+    public void onFinishedGoingToSleep_pendingAuthWithSecondFactorEnabled_setsUserAuthWithFp() {
+        when(mUpdateMonitor.isGoingToSleep()).thenReturn(true);
+        mBiometricUnlockController.mWakefulnessObserver.onFinishedGoingToSleep();
+        verify(mHandler, never()).post(any());
+
+        ArgumentCaptor<Runnable> captor = ArgumentCaptor.forClass(Runnable.class);
+        mBiometricUnlockController.onBiometricAuthenticated(1,
+                BiometricSourceType.FINGERPRINT, true);
+        mBiometricUnlockController.mWakefulnessObserver.onFinishedGoingToSleep();
+        verify(mHandler).post(captor.capture());
+        captor.getValue().run();
+        verify(mUpdateMonitor).setUserAuthenticatedWithFingerprint(1, true);
+    }
+
     @Test
     public void onFPFailureNoHaptics_notInteractive_showLockScreen() {
         // GIVEN no vibrator and device is not interactive
@@ -555,7 +621,8 @@ public class BiometricsUnlockControllerTest extends SysuiTestCase {
                 .thenReturn(PowerManager.WAKE_REASON_POWER_BUTTON);
         givenDreamingLocked();
         when(mPowerManager.isInteractive()).thenReturn(true);
-        mBiometricUnlockController.startWakeAndUnlock(BiometricSourceType.FINGERPRINT, true);
+        mBiometricUnlockController.startWakeAndUnlock(BiometricSourceType.FINGERPRINT, true,
+                Disabled);
         verify(mKeyguardViewMediator).onWakeAndUnlocking(true);
         // Ensure that the power hasn't been told to wake up yet.
         verify(mPowerManager, never()).wakeUp(anyLong(), anyInt(), anyString());
diff --git a/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java b/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java
index 089e3048a..390293575 100644
--- a/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java
+++ b/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java
@@ -43,9 +43,12 @@ import static com.android.internal.widget.LockPatternUtils.StrongAuthTracker.STR
 import static com.android.internal.widget.LockPatternUtils.StrongAuthTracker.STRONG_AUTH_REQUIRED_AFTER_DPM_LOCK_NOW;
 import static com.android.internal.widget.LockPatternUtils.StrongAuthTracker.STRONG_AUTH_REQUIRED_AFTER_LOCKOUT;
 import static com.android.internal.widget.LockPatternUtils.StrongAuthTracker.STRONG_AUTH_REQUIRED_AFTER_USER_LOCKDOWN;
+import static com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus.Disabled;
+import static com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus.Enabled;
 import static com.android.systemui.Flags.simPinBouncerReset;
 import static com.android.systemui.Flags.simPinUseSlotId;
 import static com.android.systemui.statusbar.policy.DevicePostureController.DEVICE_POSTURE_OPENED;
+import static java.util.Objects.requireNonNull;
 
 import android.annotation.AnyThread;
 import android.annotation.MainThread;
@@ -116,6 +119,7 @@ import com.android.internal.logging.InstanceId;
 import com.android.internal.logging.UiEventLogger;
 import com.android.internal.util.LatencyTracker;
 import com.android.internal.widget.LockPatternUtils;
+import com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus;
 import com.android.keyguard.logging.KeyguardUpdateMonitorLogger;
 import com.android.keyguard.logging.SimLogger;
 import com.android.settingslib.Utils;
@@ -182,6 +186,7 @@ import java.util.Optional;
 import java.util.Set;
 import java.util.TimeZone;
 import java.util.concurrent.Executor;
+import java.util.function.BooleanSupplier;
 import java.util.stream.Collectors;
 
 import javax.inject.Inject;
@@ -271,6 +276,9 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener, Dumpab
     public static final int BIOMETRIC_HELP_FACE_NOT_RECOGNIZED = -2;
     public static final int BIOMETRIC_HELP_FACE_NOT_AVAILABLE = -3;
 
+    // TODO: As there were other race conditions found in this code, this should be investigated.
+    //  How was this time determined? What would cause there to be a timeout? Are we sure that
+    //  no result can be returned after the timeout handler has executed?
     /**
      * If no cancel signal has been received after this amount of time, set the biometric running
      * state to stopped to allow Keyguard to retry authentication.
@@ -881,40 +889,110 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener, Dumpab
         Trace.endSection();
     }
 
+    public static final String BSF_TAG = "BiometricSecondFactor";
+
     @VisibleForTesting
     public void onFingerprintAuthenticated(int userId, boolean isStrongBiometric) {
         Assert.isMainThread();
         Trace.beginSection("KeyGuardUpdateMonitor#onFingerPrintAuthenticated");
         mUserFingerprintAuthenticated.put(userId,
                 new BiometricAuthenticated(true, isStrongBiometric));
+
+        final boolean isBiometricSecondFactorEnabled = mLockPatternUtils.isBiometricSecondFactorEnabled(userId)
+                && !getUserCanSkipBouncer(userId);
+
         // Update/refresh trust state only if user can skip bouncer
         if (getUserCanSkipBouncer(userId)) {
+            // Called by pendingSecondFactorAction after second factor succeeds.
             mTrustManager.unlockedByBiometricForUser(userId, FINGERPRINT);
         }
         // Don't send cancel if authentication succeeds
         mFingerprintCancelSignal = null;
         mLogger.logFingerprintSuccess(userId, isStrongBiometric);
+
+        // Assuming we aren't in BIOMETRIC_STATE_STOPPED (refer to comments on
+        // DEFAULT_CANCEL_SIGNAL_TIMEOUT), this call is essentially a no-op. Actually, before my
+        // fixes to #updateFingerprintListeningState, it was a bug to call this here. It would
+        // probably make sense to call this after state is set to BIOMETRIC_STATE_STOPPED, which
+        // happens after we return. No point removing as it is called by BiometricUnlockController
+        // anyway.
         updateFingerprintListeningState(BIOMETRIC_ACTION_UPDATE);
+
+        SecondFactorStatus secondFactorStatus = isBiometricSecondFactorEnabled ? Enabled : Disabled;
         for (int i = 0; i < mCallbacks.size(); i++) {
             KeyguardUpdateMonitorCallback cb = mCallbacks.get(i).get();
             if (cb != null) {
-                cb.onBiometricAuthenticated(userId, FINGERPRINT,
-                        isStrongBiometric);
+                // secondFactorStatus param is redundant but will help prevent silent bugs
+                // being introduced by future code and is now something that all callbacks should
+                // consider.
+                cb.onBiometricAuthenticated(userId, FINGERPRINT, isStrongBiometric,
+                        secondFactorStatus);
             }
         }
 
+        // Upstream doesn't check getUserCanSkipBouncer, so not going to check SecondFactorStatus
+        // == Disabled.
         mHandler.sendMessageDelayed(mHandler.obtainMessage(MSG_BIOMETRIC_AUTHENTICATION_CONTINUE),
                 FINGERPRINT_CONTINUE_DELAY_MS);
 
         // Only authenticate fingerprint once when assistant is visible
         mAssistantVisible = false;
 
-        // Report unlock with strong or non-strong biometric
-        reportSuccessfulBiometricUnlock(isStrongBiometric, userId);
+        if (getUserCanSkipBouncer(userId)) {
+            // This can enable non-strong biometrics, which shouldn't happen until after the second
+            // factor succeeds. Called by pendingSecondFactorAction below after second factor succeeds.
+            reportSuccessfulBiometricUnlock(isStrongBiometric, userId);
+        }
+
+        if (isBiometricSecondFactorEnabled) {
+            Log.d(BSF_TAG, "onFingerprintAuthenticated, userId: " + userId + ", isStrongBiometric: " + isStrongBiometric);
+            // pendingSecondFactorAction lambda is executed after second factor PIN is verified.
+            // Returning false from it stops the unlock sequence.
+            BooleanSupplier pendingSecondFactorAction = () -> {
+                FingerprintManager fpm = requireNonNull(mFpm, "fingerprintManager");
+                int res = fpm.addPendingAuthTokenToKeyStore(userId);
+
+                switch (res) {
+                    case FingerprintManager.SUCCESS:
+                        break;
+                    case FingerprintManager.ERROR_UNABLE_TO_ADD_AUTH_TOKEN_TO_KEYSTORE:
+                        Log.e(BSF_TAG, "addPendingAuthTokenToKeyStore returned ERROR_UNABLE_TO_ADD_AUTH_TOKEN_TO_KEYSTORE");
+                        return false;
+                    case FingerprintManager.ERROR_NO_PENDING_AUTH_TOKEN:
+                        Log.e(BSF_TAG, "addPendingAuthTokenToKeyStore returned ERROR_NO_PENDING_AUTH_TOKEN");
+                        return false;
+                    default:
+                        Log.e(BSF_TAG, "addPendingAuthTokenToKeyStore returned unknown value: " + res);
+                        return false;
+                }
+
+                mTrustManager.unlockedByBiometricForUser(userId, FINGERPRINT);
+                reportSuccessfulBiometricUnlock(isStrongBiometric, userId);
+                return true;
+            };
+            Assert.isMainThread();
+            pendingSecondFactorActions.put(userId, pendingSecondFactorAction);
+            Log.d(BSF_TAG, "onFingerprintAuthenticated: added pendingSecondFactorAction");
+        }
 
         Trace.endSection();
     }
 
+    private static final SparseArray<BooleanSupplier> pendingSecondFactorActions =
+            new SparseArray<>();
+
+    @Nullable
+    static BooleanSupplier getAndRemovePendingSecondFactorAction(int userId) {
+        Assert.isMainThread();
+
+        BooleanSupplier res = pendingSecondFactorActions.get(userId);
+        if (res == null) {
+            return null;
+        }
+        pendingSecondFactorActions.remove(userId);
+        return res;
+    }
+
     private void reportSuccessfulBiometricUnlock(boolean isStrongBiometric, int userId) {
         mBackgroundExecutor.execute(
                 () -> {
@@ -1171,17 +1249,25 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener, Dumpab
         for (int i = 0; i < mCallbacks.size(); i++) {
             KeyguardUpdateMonitorCallback cb = mCallbacks.get(i).get();
             if (cb != null) {
+                SecondFactorStatus secondFactorStatus =
+                        mLockPatternUtils.isBiometricSecondFactorEnabled(userId) ?
+                                Enabled : Disabled;
                 cb.onBiometricAuthenticated(userId,
                         FACE,
-                        isStrongBiometric);
+                        isStrongBiometric,
+                        secondFactorStatus);
             }
         }
 
         // Only authenticate face once when assistant is visible
         mAssistantVisible = false;
 
-        // Report unlock with strong or non-strong biometric
-        reportSuccessfulBiometricUnlock(isStrongBiometric, userId);
+        if (getUserCanSkipBouncer(userId)) {
+            // This can enable non-strong biometrics, which shouldn't happen until after the second
+            // factor succeeds. Called by LockPatternUtils#reportSuccessfulPasswordAttempt after
+            // second factor succeeds.
+            reportSuccessfulBiometricUnlock(isStrongBiometric, userId);
+        }
 
         Trace.endSection();
     }
@@ -1433,9 +1519,14 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener, Dumpab
     }
 
     /**
-     * Returns whether the user is unlocked with biometrics.
+     * Returns whether the user is unlocked with biometrics. If biometric second factor is enabled
+     * for the user then false is always returned.
      */
     public boolean getUserUnlockedWithBiometric(int userId) {
+        if (mLockPatternUtils.isBiometricSecondFactorEnabled(userId)) {
+            return false;
+        }
+
         BiometricAuthenticated fingerprint = mUserFingerprintAuthenticated.get(userId);
         boolean fingerprintAllowed = fingerprint != null && fingerprint.mAuthenticated
                 && isUnlockingWithBiometricAllowedSafe(fingerprint.mIsStrongBiometric);
@@ -1444,6 +1535,14 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener, Dumpab
         return fingerprintAllowed || unlockedByFace;
     }
 
+    public boolean getUserAuthenticatedWithFingerprint(int userId) {
+        return mUserFingerprintAuthenticated.contains(userId);
+    }
+
+    public void setUserAuthenticatedWithFingerprint(int userId, boolean isStrongBiometric) {
+        mUserFingerprintAuthenticated.put(userId,
+                new BiometricAuthenticated(true, isStrongBiometric));
+    }
 
     /**
      * Returns whether the user is unlocked with face.
@@ -1457,9 +1556,14 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener, Dumpab
 
     /**
      * Returns whether the user is unlocked with a biometric that is currently bypassing
-     * the lock screen.
+     * the lock screen. If biometric second factor is enabled for the user then false is always
+     * returned.
      */
     public boolean getUserUnlockedWithBiometricAndIsBypassing(int userId) {
+        if (mLockPatternUtils.isBiometricSecondFactorEnabled(userId)) {
+            return false;
+        }
+
         BiometricAuthenticated fingerprint = mUserFingerprintAuthenticated.get(userId);
         // fingerprint always bypasses
         boolean fingerprintAllowed = fingerprint != null && fingerprint.mAuthenticated
@@ -1545,7 +1649,9 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener, Dumpab
         // STRONG_AUTH_REQUIRED_AFTER_LOCKOUT which is the same as mFingerprintLockedOutPermanent;
         // however the strong auth tracker does not include the temporary lockout
         // mFingerprintLockedOut.
-        if (!mStrongAuthTracker.isUnlockingWithBiometricAllowed(isStrongBiometric)) {
+        int userId = mSelectedUserInteractor.getSelectedUserId();
+        if (!mStrongAuthTracker.isUnlockingWithBiometricAllowed(isStrongBiometric)
+                || getUserWaitingForStrongAuthUpdateAfterSecondFactorLockout(userId)) {
             return false;
         }
         boolean isFaceLockedOut =
@@ -2047,6 +2153,9 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener, Dumpab
 
         @Override
         public void onStrongAuthRequiredChanged(int userId) {
+            if ((getStrongAuthForUser(userId) & STRONG_AUTH_REQUIRED_AFTER_LOCKOUT) != 0) {
+                setUserWaitingForStrongAuthUpdateAfterSecondFactorLockout(userId, false);
+            }
             notifyStrongAuthAllowedChanged(userId);
         }
 
@@ -2967,6 +3076,18 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener, Dumpab
                 && !mUserHasTrust.get(mSelectedUserInteractor.getSelectedUserId(), false);
     }
 
+    private final SparseBooleanArray mUserWaitingForStrongAuthUpdateAfterSecondFactorLockout =
+            new SparseBooleanArray();
+
+    protected void setUserWaitingForStrongAuthUpdateAfterSecondFactorLockout(int userId,
+            boolean value) {
+        mUserWaitingForStrongAuthUpdateAfterSecondFactorLockout.put(userId, value);
+    }
+
+    private boolean getUserWaitingForStrongAuthUpdateAfterSecondFactorLockout(int userId) {
+        return mUserWaitingForStrongAuthUpdateAfterSecondFactorLockout.get(userId);
+    }
+
     @VisibleForTesting
     protected boolean shouldListenForFingerprint(boolean isUdfps) {
         final int user = mSelectedUserInteractor.getSelectedUserId();
@@ -2999,9 +3120,19 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener, Dumpab
                         && mIsSystemUser
                         && biometricEnabledForUser
                         && !isUserInLockdown(user);
-        final boolean strongerAuthRequired = !isUnlockingWithFingerprintAllowed();
+
+        final boolean strongerAuthRequired = !isUnlockingWithFingerprintAllowedSafe();
+        final boolean fingerprintAuthenticated = getUserAuthenticatedWithFingerprint(user);
+        final boolean biometricSecondFactorEnabled = mLockPatternUtils
+                .isBiometricSecondFactorEnabled(user);
         final boolean shouldListenBouncerState =
-                !strongerAuthRequired || !isPrimaryBouncerShowingOrWillBeShowing();
+                (!strongerAuthRequired || !isPrimaryBouncerShowingOrWillBeShowing())
+                        // Disable fp on second factor bouncer. Can't truly know if we are on second
+                        // factor bouncer because it depends on isUnlockingWithFingerprintAllowed,
+                        // which can change at any given point. Worst case we end up disabled on
+                        // primary bouncer, but even in that case the first leg of
+                        // shouldListenBouncerState will usually be false anyway.
+                        && !(fingerprintAuthenticated && biometricSecondFactorEnabled);
 
         final boolean shouldListenUdfpsState = !isUdfps
                 || (!userCanSkipBouncer
@@ -3853,12 +3984,27 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener, Dumpab
         clearFingerprintRecognized(unlockedUser);
     }
 
-    private void clearFingerprintRecognized(int unlockedUser) {
+    protected void clearFingerprintRecognized(int unlockedUser) {
         Assert.isMainThread();
         mUserFingerprintAuthenticated.clear();
         mTrustManager.clearAllBiometricRecognized(FINGERPRINT, unlockedUser);
         mLogger.d("clearFingerprintRecognized");
 
+        if (mFpm != null) {
+            mFpm.clearPendingAuthTokens();
+        }
+        Assert.isMainThread();
+        final int psfaSize = pendingSecondFactorActions.size();
+        if (psfaSize != 0) {
+            int[] keys = new int[psfaSize];
+            for (int i = 0; i < psfaSize; ++i) {
+                keys[i] = pendingSecondFactorActions.keyAt(i);
+            }
+            pendingSecondFactorActions.clear();
+            Log.d(BSF_TAG, "clearFingerprintRecognized: cleared pendingActions, "
+                    + "past keys: " + Arrays.toString(keys) + ", unlockedUser: " + unlockedUser);
+        }
+
         for (int i = 0; i < mCallbacks.size(); i++) {
             KeyguardUpdateMonitorCallback cb = mCallbacks.get(i).get();
             if (cb != null) {
diff --git a/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitorCallback.java b/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitorCallback.java
index 872942c49..640edd951 100644
--- a/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitorCallback.java
+++ b/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitorCallback.java
@@ -215,13 +215,19 @@ public class KeyguardUpdateMonitorCallback {
      */
     public void onBiometricAuthFailed(BiometricSourceType biometricSourceType) { }
 
+    // Use custom type instead of boolean in case upstream overloads/extends this callback
+    // themselves.
+    public enum SecondFactorStatus {
+        Enabled, Disabled
+    }
+
     /**
      * Called when a biometric is authenticated.
      * @param userId the user id for which the biometric sample was authenticated
      * @param biometricSourceType
      */
     public void onBiometricAuthenticated(int userId, BiometricSourceType biometricSourceType,
-            boolean isStrongBiometric) { }
+            boolean isStrongBiometric, SecondFactorStatus secondFactorStatus) { }
 
     /**
      * Called when a biometric is detected but not successfully authenticated.
diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/AuthRippleController.kt b/packages/SystemUI/src/com/android/systemui/biometrics/AuthRippleController.kt
index f6cc72431..f29553b61 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/AuthRippleController.kt
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/AuthRippleController.kt
@@ -27,6 +27,7 @@ import android.util.DisplayMetrics
 import androidx.annotation.VisibleForTesting
 import androidx.lifecycle.repeatOnLifecycle
 import com.android.app.animation.Interpolators
+import com.android.internal.widget.LockPatternUtils
 import com.android.keyguard.KeyguardUpdateMonitor
 import com.android.keyguard.KeyguardUpdateMonitorCallback
 import com.android.keyguard.logging.KeyguardLogger
@@ -52,6 +53,7 @@ import com.android.systemui.statusbar.commandline.CommandRegistry
 import com.android.systemui.statusbar.phone.BiometricUnlockController
 import com.android.systemui.statusbar.policy.ConfigurationController
 import com.android.systemui.statusbar.policy.KeyguardStateController
+import com.android.systemui.user.domain.interactor.SelectedUserInteractor
 import com.android.systemui.util.ViewController
 import java.io.PrintWriter
 import javax.inject.Inject
@@ -85,6 +87,7 @@ constructor(
     private val lightRevealScrim: LightRevealScrim,
     private val authRippleInteractor: AuthRippleInteractor,
     private val facePropertyRepository: FacePropertyRepository,
+    private val selectedUserInteractor: SelectedUserInteractor,
     rippleView: AuthRippleView?,
 ) :
     ViewController<AuthRippleView>(rippleView),
@@ -147,8 +150,15 @@ constructor(
 
     private fun showUnlockRippleInternal(biometricSourceType: BiometricSourceType) {
         val keyguardNotShowing = !keyguardStateController.isShowing
+        // TODO: When adding this I noticed there is a bug that prevents unlock ripple from
+        //  displaying when a user adds and then deletes a second factor. It occurs even when
+        //  setting a second factor and then deleting it before going to the lockscreen. It occurs
+        //  when changing primary credential also, so we probably didn't introduce it.
+        val secondFactorEnabled = LockPatternUtils(sysuiContext).isBiometricSecondFactorEnabled(
+            selectedUserInteractor.getSelectedUserId())
         val unlockNotAllowed =
-            !keyguardUpdateMonitor.isUnlockingWithBiometricAllowed(biometricSourceType)
+            !keyguardUpdateMonitor.isUnlockingWithBiometricAllowedSafe(biometricSourceType) ||
+                secondFactorEnabled
         if (keyguardNotShowing || unlockNotAllowed) {
             logger.notShowingUnlockRipple(keyguardNotShowing, unlockNotAllowed)
             return
@@ -291,9 +301,16 @@ constructor(
                 userId: Int,
                 biometricSourceType: BiometricSourceType,
                 isStrongBiometric: Boolean,
+                secondFactorStatus: SecondFactorStatus,
             ) {
                 if (biometricSourceType == BiometricSourceType.FINGERPRINT) {
-                    mView.fadeDwellRipple()
+                    if (secondFactorStatus == SecondFactorStatus.Enabled) {
+                        // Do same as in #onBiometricAuthFailed as want to keep second factor UI
+                        // in sync with going from 3rd UDFPS failure to bouncer.
+                        mView.retractDwellRipple()
+                    } else {
+                        mView.fadeDwellRipple()
+                    }
                 }
             }
 
diff --git a/packages/SystemUI/src/com/android/systemui/bouncer/ui/binder/KeyguardBouncerViewBinder.kt b/packages/SystemUI/src/com/android/systemui/bouncer/ui/binder/KeyguardBouncerViewBinder.kt
index 9fe942cd5..f96fd158a 100644
--- a/packages/SystemUI/src/com/android/systemui/bouncer/ui/binder/KeyguardBouncerViewBinder.kt
+++ b/packages/SystemUI/src/com/android/systemui/bouncer/ui/binder/KeyguardBouncerViewBinder.kt
@@ -262,11 +262,22 @@ object KeyguardBouncerViewBinder {
 
                     launch {
                         viewModel.bouncerRequestedWhenAlreadyShowing.collect {
-                            // This will be a no-op except for when fingerprint authenticated
-                            // while the primary bouncer is showing. It would seem this is only
-                            // possible for non-UDFPS devices, but there might be edge cases or bugs
-                            // where UDFPS is displayed on top of primary bouncer.
-                            securityContainerController.showPrimarySecurityScreen(false)
+                            // We get here when fp authenticated while the primary bouncer is
+                            // showing. If KUM#getUserCanSkipBouncer is true then we should dismiss
+                            // immediately. This same dismissal attempt is also done when showing
+                            // the bouncer initially as part of PrimaryBouncerInteractor#show.
+                            val dismissed = securityContainerController.dismiss(
+                                selectedUserInteractor.getSelectedUserId()
+                            )
+                            if (!dismissed) {
+                                // This will be a no-op except for when fingerprint authenticated
+                                // while the primary bouncer is showing (as well as SIM PIN/PUK,
+                                // which is discussed in SBKVM#showPrimaryBouncer). It would seem
+                                // this is only possible for non-UDFPS devices, but there might be
+                                // edge cases or bugs where UDFPS is displayed on top of primary
+                                // bouncer.
+                                securityContainerController.showPrimarySecurityScreen(false)
+                            }
                             viewModel.notifyBouncerRequestedWhenAlreadyShowingHandled()
                         }
                     }
diff --git a/packages/SystemUI/src/com/android/systemui/classifier/FalsingCollectorImpl.java b/packages/SystemUI/src/com/android/systemui/classifier/FalsingCollectorImpl.java
index 962ab998d..ab9c5b340 100644
--- a/packages/SystemUI/src/com/android/systemui/classifier/FalsingCollectorImpl.java
+++ b/packages/SystemUI/src/com/android/systemui/classifier/FalsingCollectorImpl.java
@@ -130,7 +130,8 @@ class FalsingCollectorImpl implements FalsingCollector {
                 @Override
                 public void onBiometricAuthenticated(int userId,
                         BiometricSourceType biometricSourceType,
-                        boolean isStrongBiometric) {
+                        boolean isStrongBiometric,
+                        SecondFactorStatus secondFactorStatus) {
                     if (userId == mUserInteractor.get().getSelectedUserId()
                             && biometricSourceType == BiometricSourceType.FACE) {
                         mFalsingDataProvider.setJustUnlockedWithFace(true);
diff --git a/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java b/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
index b0c82063a..3a2d70b9d 100644
--- a/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
+++ b/packages/SystemUI/src/com/android/systemui/keyguard/KeyguardViewMediator.java
@@ -41,6 +41,7 @@ import static com.android.internal.widget.LockPatternUtils.StrongAuthTracker.STR
 import static com.android.internal.widget.LockPatternUtils.StrongAuthTracker.STRONG_AUTH_REQUIRED_AFTER_TIMEOUT;
 import static com.android.internal.widget.LockPatternUtils.StrongAuthTracker.STRONG_AUTH_REQUIRED_AFTER_USER_LOCKDOWN;
 import static com.android.internal.widget.LockPatternUtils.StrongAuthTracker.STRONG_AUTH_REQUIRED_FOR_UNATTENDED_UPDATE;
+import static com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus.Disabled;
 import static com.android.systemui.DejankUtils.whitelistIpcs;
 import static com.android.systemui.Flags.notifyPowerManagerUserActivityBackground;
 import static com.android.systemui.Flags.relockWithPowerButtonImmediately;
@@ -867,13 +868,13 @@ public class KeyguardViewMediator implements CoreStartable, Dumpable,
 
         @Override
         public void onBiometricAuthenticated(int userId, BiometricSourceType biometricSourceType,
-                boolean isStrongBiometric) {
+                boolean isStrongBiometric, SecondFactorStatus secondFactorStatus) {
             if (userId != mSelectedUserInteractor.getSelectedUserId()) {
                 Log.w(TAG, "onBiometricAuthenticated() invoked for userId: " + userId + ", current "
                         + "userId: " + mSelectedUserInteractor.getSelectedUserId());
                 return;
             }
-            if (mLockPatternUtils.isSecure(userId)) {
+            if (mLockPatternUtils.isSecure(userId) && secondFactorStatus == Disabled) {
                 mLockPatternUtils.getDevicePolicyManager().reportSuccessfulBiometricAttempt(
                         userId);
             }
diff --git a/packages/SystemUI/src/com/android/systemui/keyguard/data/repository/DeviceEntryFingerprintAuthRepository.kt b/packages/SystemUI/src/com/android/systemui/keyguard/data/repository/DeviceEntryFingerprintAuthRepository.kt
index 4d999df69..998c3bb21 100644
--- a/packages/SystemUI/src/com/android/systemui/keyguard/data/repository/DeviceEntryFingerprintAuthRepository.kt
+++ b/packages/SystemUI/src/com/android/systemui/keyguard/data/repository/DeviceEntryFingerprintAuthRepository.kt
@@ -211,7 +211,11 @@ constructor(
                                 userId: Int,
                                 biometricSourceType: BiometricSourceType,
                                 isStrongBiometric: Boolean,
+                                secondFactorStatus: SecondFactorStatus,
                             ) {
+                                if (secondFactorStatus == SecondFactorStatus.Enabled) {
+                                    return
+                                }
                                 sendUpdateIfFingerprint(
                                     biometricSourceType,
                                     SuccessFingerprintAuthenticationStatus(
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/KeyguardIndicationController.java b/packages/SystemUI/src/com/android/systemui/statusbar/KeyguardIndicationController.java
index 17c9e2429..2d4438a79 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/KeyguardIndicationController.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/KeyguardIndicationController.java
@@ -1503,8 +1503,9 @@ public class KeyguardIndicationController {
 
         @Override
         public void onBiometricAuthenticated(int userId, BiometricSourceType biometricSourceType,
-                boolean isStrongBiometric) {
-            super.onBiometricAuthenticated(userId, biometricSourceType, isStrongBiometric);
+                boolean isStrongBiometric, SecondFactorStatus secondFactorStatus) {
+            super.onBiometricAuthenticated(userId, biometricSourceType, isStrongBiometric,
+                    secondFactorStatus);
             hideBiometricMessage();
             if (biometricSourceType == FACE) {
                 mFaceAcquiredMessageDeferral.reset();
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/phone/BiometricUnlockController.java b/packages/SystemUI/src/com/android/systemui/statusbar/phone/BiometricUnlockController.java
index 8658115ba..692a62e68 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/phone/BiometricUnlockController.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/phone/BiometricUnlockController.java
@@ -17,6 +17,7 @@
 package com.android.systemui.statusbar.phone;
 
 import static android.app.StatusBarManager.SESSION_KEYGUARD;
+import static com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus.Disabled;
 
 import android.annotation.IntDef;
 import android.content.res.Resources;
@@ -195,12 +196,14 @@ public class BiometricUnlockController extends KeyguardUpdateMonitorCallback imp
         public final int userId;
         public final BiometricSourceType biometricSourceType;
         public final boolean isStrongBiometric;
+        public final SecondFactorStatus secondFactorStatus;
 
         PendingAuthenticated(int userId, BiometricSourceType biometricSourceType,
-                boolean isStrongBiometric) {
+                boolean isStrongBiometric, SecondFactorStatus secondFactorStatus) {
             this.userId = userId;
             this.biometricSourceType = biometricSourceType;
             this.isStrongBiometric = isStrongBiometric;
+            this.secondFactorStatus = secondFactorStatus;
         }
     }
 
@@ -423,16 +426,22 @@ public class BiometricUnlockController extends KeyguardUpdateMonitorCallback imp
         Trace.endSection();
     }
 
-    @Override
+    // Keep this overload to reduce amount of changes required in upstream tests.
     public void onBiometricAuthenticated(int userId, BiometricSourceType biometricSourceType,
             boolean isStrongBiometric) {
+        onBiometricAuthenticated(userId, biometricSourceType, isStrongBiometric, Disabled);
+    }
+
+    @Override
+    public void onBiometricAuthenticated(int userId, BiometricSourceType biometricSourceType,
+            boolean isStrongBiometric, SecondFactorStatus secondFactorStatus) {
         Trace.beginSection("BiometricUnlockController#onBiometricUnlocked");
         if (mUpdateMonitor.isGoingToSleep()) {
             mLogger.deferringAuthenticationDueToSleep(userId,
                     biometricSourceType,
                     mPendingAuthenticated != null);
             mPendingAuthenticated = new PendingAuthenticated(userId, biometricSourceType,
-                    isStrongBiometric);
+                    isStrongBiometric, secondFactorStatus);
             Trace.endSection();
             return;
         }
@@ -445,10 +454,10 @@ public class BiometricUnlockController extends KeyguardUpdateMonitorCallback imp
         boolean unlockAllowed =
                 mKeyguardStateController.isOccluded()
                         || mKeyguardBypassController.onBiometricAuthenticated(
-                                biometricSourceType, isStrongBiometric);
+                                biometricSourceType, isStrongBiometric, secondFactorStatus);
         if (unlockAllowed) {
             mKeyguardViewMediator.userActivity();
-            startWakeAndUnlock(biometricSourceType, isStrongBiometric);
+            startWakeAndUnlock(biometricSourceType, isStrongBiometric, secondFactorStatus);
         } else {
             mLogger.d("onBiometricUnlocked aborted by bypass controller");
         }
@@ -459,10 +468,11 @@ public class BiometricUnlockController extends KeyguardUpdateMonitorCallback imp
      * Wake and unlock the device in response to successful authentication using biometrics.
      * @param biometricSourceType Biometric source that was used to authenticate.
      * @param isStrongBiometric
+     * @param secondFactorStatus
      */
     public void startWakeAndUnlock(BiometricSourceType biometricSourceType,
-                                   boolean isStrongBiometric) {
-        int mode = calculateMode(biometricSourceType, isStrongBiometric);
+            boolean isStrongBiometric, SecondFactorStatus secondFactorStatus) {
+        int mode = calculateMode(biometricSourceType, isStrongBiometric, secondFactorStatus);
         if (mode == MODE_WAKE_AND_UNLOCK
                 || mode == MODE_WAKE_AND_UNLOCK_PULSING || mode == MODE_UNLOCK_COLLAPSING
                 || mode == MODE_WAKE_AND_UNLOCK_FROM_DREAM || mode == MODE_DISMISS_BOUNCER) {
@@ -588,29 +598,41 @@ public class BiometricUnlockController extends KeyguardUpdateMonitorCallback imp
     }
 
     private @WakeAndUnlockMode int calculateMode(BiometricSourceType biometricSourceType,
-            boolean isStrongBiometric) {
+            boolean isStrongBiometric, SecondFactorStatus secondFactorStatus) {
         if (biometricSourceType == BiometricSourceType.FACE
                 || biometricSourceType == BiometricSourceType.IRIS) {
             return calculateModeForPassiveAuth(isStrongBiometric);
         } else {
-            return calculateModeForFingerprint(isStrongBiometric);
+            return calculateModeForFingerprint(isStrongBiometric, secondFactorStatus);
         }
     }
 
-    private @WakeAndUnlockMode int calculateModeForFingerprint(boolean isStrongBiometric) {
+    private @WakeAndUnlockMode int calculateModeForFingerprint(boolean isStrongBiometric,
+            SecondFactorStatus secondFactorStatus) {
         final boolean unlockingAllowed =
-                mUpdateMonitor.isUnlockingWithBiometricAllowed(isStrongBiometric);
+                mUpdateMonitor.isUnlockingWithBiometricAllowedSafe(isStrongBiometric) &&
+                secondFactorStatus == Disabled;
         final boolean deviceInteractive = mUpdateMonitor.isDeviceInteractive();
         final boolean keyguardShowing = mKeyguardStateController.isShowing();
         final boolean deviceDreaming = mUpdateMonitor.isDreaming();
-
         logCalculateModeForFingerprint(unlockingAllowed, deviceInteractive,
                 keyguardShowing, deviceDreaming, isStrongBiometric);
         if (!deviceInteractive) {
             if (!keyguardShowing && !mScreenOffAnimationController.isKeyguardShowDelayed()) {
+                // The purpose of this code is not entirely clear. It appears that
+                // !deviceInteractive && !keyguardShowing is only true when the primary credential
+                // is None (not Swipe). When primary is None, it's not possible for the user to have
+                // a fingerprint registered. Perhaps there is a very small window where the
+                // screen is off, keyguard has yet to show (or be set as showing delayed) and the
+                // fingerprint is being listened for.
+
+                // !keyguardShowing implies mKeyguardStateController.isUnlocked(), so this
+                // conditional is always true. Everything is running in the same thread so not
+                // possible for this implication to be invalidated externally.
                 if (mKeyguardStateController.isUnlocked()) {
                     return MODE_WAKE_AND_UNLOCK;
                 }
+                // This is unreachable.
                 return MODE_ONLY_WAKE;
             } else if (mDozeScrimController.isPulsing() && unlockingAllowed) {
                 return MODE_WAKE_AND_UNLOCK_PULSING;
@@ -628,7 +650,13 @@ public class BiometricUnlockController extends KeyguardUpdateMonitorCallback imp
                 return MODE_DISMISS_BOUNCER;
             } else if (unlockingAllowed) {
                 return MODE_UNLOCK_COLLAPSING;
-            } else if (!mKeyguardViewController.isBouncerShowing()) {
+            } else {
+                // As of 15 QPR1, the alternate bouncer is displayed when switching to a user that
+                // has fingerprint unlock enabled. This makes upstream's conditional on
+                // !mKeyguardViewController.isBouncerShowing() incorrect, as they should only be
+                // checking the primary bouncer not showing. We remove the conditional entirely, as
+                // second factor auth requires MODE_SHOW_BOUNCER even when the primary is showing
+                // for non-UDFPS devices.
                 return MODE_SHOW_BOUNCER;
             }
         }
@@ -659,7 +687,7 @@ public class BiometricUnlockController extends KeyguardUpdateMonitorCallback imp
         final boolean deviceInteractive = mUpdateMonitor.isDeviceInteractive();
         final boolean isKeyguardShowing = mKeyguardStateController.isShowing();
         final boolean unlockingAllowed =
-                mUpdateMonitor.isUnlockingWithBiometricAllowed(isStrongBiometric);
+                mUpdateMonitor.isUnlockingWithBiometricAllowedSafe(isStrongBiometric);
         final boolean deviceDreaming = mUpdateMonitor.isDreaming();
         final boolean bypass = mKeyguardBypassController.getBypassEnabled()
                 || mAuthController.isUdfpsFingerDown();
@@ -847,9 +875,19 @@ public class BiometricUnlockController extends KeyguardUpdateMonitorCallback imp
                         mLogger.finishedGoingToSleepWithPendingAuth();
                         PendingAuthenticated pendingAuthenticated = mPendingAuthenticated;
                         // Post this to make sure it's executed after the device is fully locked.
-                        mHandler.post(() -> onBiometricAuthenticated(pendingAuthenticated.userId,
+                        mHandler.post(() -> {
+                            // Needed for SecurityMode.BiometricSecondFactorPin.
+                            if (pendingAuthenticated.biometricSourceType ==
+                                    BiometricSourceType.FINGERPRINT) {
+                                mUpdateMonitor.setUserAuthenticatedWithFingerprint(
+                                        pendingAuthenticated.userId,
+                                        pendingAuthenticated.isStrongBiometric);
+                            }
+                            onBiometricAuthenticated(pendingAuthenticated.userId,
                                 pendingAuthenticated.biometricSourceType,
-                                pendingAuthenticated.isStrongBiometric));
+                                pendingAuthenticated.isStrongBiometric,
+                                pendingAuthenticated.secondFactorStatus);
+                        });
                         mPendingAuthenticated = null;
                     }
                     Trace.endSection();
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/phone/KeyguardBypassController.kt b/packages/SystemUI/src/com/android/systemui/statusbar/phone/KeyguardBypassController.kt
index 3242149ee..fed4fb0e5 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/phone/KeyguardBypassController.kt
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/phone/KeyguardBypassController.kt
@@ -23,6 +23,7 @@ import android.hardware.biometrics.BiometricSourceType
 import android.provider.Settings
 import com.android.app.tracing.ListenersTracing.forEachTraced
 import com.android.app.tracing.coroutines.launchTraced as launch
+import com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus
 import com.android.systemui.Dumpable
 import com.android.systemui.dagger.SysUISingleton
 import com.android.systemui.dagger.qualifiers.Application
@@ -92,7 +93,8 @@ class KeyguardBypassController @Inject constructor(
      */
     private data class PendingUnlock(
         val pendingUnlockType: BiometricSourceType,
-        val isStrongBiometric: Boolean
+        val isStrongBiometric: Boolean,
+        val secondFactorStatus: SecondFactorStatus
     )
 
     lateinit var unlockController: BiometricUnlockController
@@ -182,12 +184,14 @@ class KeyguardBypassController @Inject constructor(
      */
     fun onBiometricAuthenticated(
         biometricSourceType: BiometricSourceType,
-        isStrongBiometric: Boolean
+        isStrongBiometric: Boolean,
+        secondFactorStatus: SecondFactorStatus,
     ): Boolean {
         if (biometricSourceType == BiometricSourceType.FACE && bypassEnabled) {
             val can = canBypass()
             if (!can && (isPulseExpanding || qsExpanded)) {
-                pendingUnlock = PendingUnlock(biometricSourceType, isStrongBiometric)
+                pendingUnlock = PendingUnlock(biometricSourceType, isStrongBiometric,
+                        secondFactorStatus)
             }
             return can
         }
@@ -197,9 +201,10 @@ class KeyguardBypassController @Inject constructor(
     fun maybePerformPendingUnlock() {
         if (pendingUnlock != null) {
             if (onBiometricAuthenticated(pendingUnlock!!.pendingUnlockType,
-                            pendingUnlock!!.isStrongBiometric)) {
+                            pendingUnlock!!.isStrongBiometric,
+                            pendingUnlock!!.secondFactorStatus)) {
                 unlockController.startWakeAndUnlock(pendingUnlock!!.pendingUnlockType,
-                        pendingUnlock!!.isStrongBiometric)
+                        pendingUnlock!!.isStrongBiometric, pendingUnlock!!.secondFactorStatus)
                 pendingUnlock = null
             }
         }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/phone/KeyguardStatusBarViewController.java b/packages/SystemUI/src/com/android/systemui/statusbar/phone/KeyguardStatusBarViewController.java
index 2433b78fc..f20c54ccb 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/phone/KeyguardStatusBarViewController.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/phone/KeyguardStatusBarViewController.java
@@ -19,6 +19,7 @@ package com.android.systemui.statusbar.phone;
 import static android.app.StatusBarManager.DISABLE2_SYSTEM_ICONS;
 import static android.app.StatusBarManager.DISABLE_SYSTEM_INFO;
 
+import static com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus.Disabled;
 import static com.android.systemui.Flags.updateUserSwitcherBackground;
 import static com.android.systemui.statusbar.StatusBarState.KEYGUARD;
 import static com.android.systemui.util.kotlin.JavaAdapterKt.collectFlow;
@@ -205,10 +206,15 @@ public class KeyguardStatusBarViewController extends ViewController<KeyguardStat
                 public void onBiometricAuthenticated(
                         int userId,
                         BiometricSourceType biometricSourceType,
-                        boolean isStrongBiometric) {
+                        boolean isStrongBiometric,
+                        SecondFactorStatus secondFactorStatus) {
+                    // mFirstBypassAttempt should only be true on devices that have face auth
+                    // enabled. We're trying not to handle callbacks for face auth as they can't be
+                    // tested, but in this case we will be cautious and add the check in case there
+                    // is some edge case where this conditional is true for fingerprint auth.
                     if (mFirstBypassAttempt
-                            && mKeyguardUpdateMonitor.isUnlockingWithBiometricAllowed(
-                                    isStrongBiometric)) {
+                            && mKeyguardUpdateMonitor.isUnlockingWithBiometricAllowedSafe(
+                                    isStrongBiometric) && secondFactorStatus == Disabled) {
                         mDelayShowingKeyguardStatusBar = true;
                     }
                 }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarKeyguardViewManager.java b/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarKeyguardViewManager.java
index bd1360f6e..f96569ea8 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarKeyguardViewManager.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBarKeyguardViewManager.java
@@ -818,18 +818,35 @@ public class StatusBarKeyguardViewManager implements RemoteInputController.Callb
      * dragging it and translation should be deferred {@see KeyguardBouncer#show(boolean, boolean)}
      */
     public void showPrimaryBouncer(boolean scrimmed) {
+        // TODO: SIM PIN/PUK insertion code does not call through here. Despite that, it might be
+        //  possible for a user with second factor enabled to complete a fingerprint auth just
+        //  after inserting a SIM with a PIN/PUK, and then KSCC#showSecurityScreen would be called
+        //  with SIM PIN/PUK prior to preparing the fullscreen bouncer. Need to verify that this is
+        //  not done.
+
         hideAlternateBouncer(
                 /* updateScrim= */ false,
                 // When the scene framework is on, don't ever clear the pending dismiss action from
                 /* clearDismissAction= */ !SceneContainerFlag.isEnabled());
-        if (mKeyguardStateController.isShowing() && !isBouncerShowing()) {
-            if (SceneContainerFlag.isEnabled()) {
-                mSceneInteractorLazy.get().changeScene(
-                        Scenes.Bouncer,
-                        "primary bouncer requested"
-                );
+        if (mKeyguardStateController.isShowing()) {
+            if (!isBouncerShowing()) {
+                if (SceneContainerFlag.isEnabled()) {
+                    mSceneInteractorLazy.get().changeScene(
+                            Scenes.Bouncer,
+                            "primary bouncer requested");
+                } else {
+                    mPrimaryBouncerInteractor.show(scrimmed);
+                }
             } else {
-                mPrimaryBouncerInteractor.show(scrimmed);
+                if (SceneContainerFlag.isEnabled()) {
+                    throw new IllegalStateException(
+                            "Handle SceneContainerFlag in SBKVM#showPrimaryBouncer!");
+                }
+                // This will either dismiss the Keyguard (if KUM#getUserCanSkipBouncer is true) or
+                // display the appropriate SecurityMode in KSCC. These are the same outcomes that
+                // a caller can expect when calling #showPrimaryBouncer while the bouncer is not
+                // showing.
+                mPrimaryBouncerInteractor.notifyBouncerRequestedWhenAlreadyShowing();
             }
         }
         updateStates();
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/policy/KeyguardStateControllerImpl.java b/packages/SystemUI/src/com/android/systemui/statusbar/policy/KeyguardStateControllerImpl.java
index 31cae79c6..4809d6db9 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/policy/KeyguardStateControllerImpl.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/policy/KeyguardStateControllerImpl.java
@@ -17,7 +17,7 @@
 package com.android.systemui.statusbar.policy;
 
 import static android.hardware.biometrics.BiometricSourceType.FACE;
-
+import static com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus.Disabled;
 import static com.android.systemui.flags.Flags.LOCKSCREEN_ENABLE_LANDSCAPE;
 
 import android.annotation.NonNull;
@@ -461,9 +461,10 @@ public class KeyguardStateControllerImpl implements KeyguardStateController {
 
         @Override
         public void onBiometricAuthenticated(int userId, BiometricSourceType biometricSourceType,
-                boolean isStrongBiometric) {
+                boolean isStrongBiometric, SecondFactorStatus secondFactorStatus) {
             Trace.beginSection("KeyguardUpdateMonitorCallback#onBiometricAuthenticated");
-            if (mKeyguardUpdateMonitor.isUnlockingWithBiometricAllowed(isStrongBiometric)) {
+            if (mKeyguardUpdateMonitor.isUnlockingWithBiometricAllowedSafe(isStrongBiometric) &&
+                    secondFactorStatus == Disabled) {
                 update(false /* updateAlways */);
             }
             Trace.endSection();
diff --git a/packages/SystemUI/tests/src/com/android/keyguard/KeyguardUpdateMonitorTest.java b/packages/SystemUI/tests/src/com/android/keyguard/KeyguardUpdateMonitorTest.java
index 7319a03ec..8e77eb777 100644
--- a/packages/SystemUI/tests/src/com/android/keyguard/KeyguardUpdateMonitorTest.java
+++ b/packages/SystemUI/tests/src/com/android/keyguard/KeyguardUpdateMonitorTest.java
@@ -128,6 +128,7 @@ import com.android.internal.util.LatencyTracker;
 import com.android.internal.widget.ILockSettings;
 import com.android.internal.widget.LockPatternUtils;
 import com.android.keyguard.KeyguardUpdateMonitor.BiometricAuthenticated;
+import com.android.keyguard.KeyguardUpdateMonitorCallback.SecondFactorStatus;
 import com.android.keyguard.logging.KeyguardUpdateMonitorLogger;
 import com.android.keyguard.logging.SimLogger;
 import com.android.settingslib.fuelgauge.BatteryStatus;
@@ -391,6 +392,8 @@ public class KeyguardUpdateMonitorTest extends SysuiTestCase {
         setupBiometrics(mKeyguardUpdateMonitor);
         mKeyguardUpdateMonitor.setFaceAuthInteractor(mFaceAuthInteractor);
         verify(mFaceAuthInteractor).registerListener(mFaceAuthenticationListener.capture());
+
+        when(mLockPatternUtils.isBiometricSecondFactorEnabled(anyInt())).thenReturn(false);
     }
 
     private void setupBiometrics(KeyguardUpdateMonitor keyguardUpdateMonitor)
@@ -425,8 +428,6 @@ public class KeyguardUpdateMonitorTest extends SysuiTestCase {
         verify(mFingerprintManager).addAuthenticatorsRegisteredCallback(
                 fingerprintCaptor.capture());
         mFingerprintAuthenticatorsRegisteredCallback = fingerprintCaptor.getValue();
-        mFingerprintAuthenticatorsRegisteredCallback
-                .onAllAuthenticatorsRegistered(mFingerprintSensorProperties);
     }
 
     private void setupFingerprintAuth(boolean isClass3) throws RemoteException {
@@ -881,7 +882,7 @@ public class KeyguardUpdateMonitorTest extends SysuiTestCase {
         verify(mTestCallback).onBiometricDetected(
                 eq(0), eq(BiometricSourceType.FINGERPRINT), eq(true));
         verify(mTestCallback, never()).onBiometricAuthenticated(
-                anyInt(), any(), anyBoolean());
+                anyInt(), any(), anyBoolean(), any(SecondFactorStatus.class));
     }
 
     @Test
@@ -898,7 +899,7 @@ public class KeyguardUpdateMonitorTest extends SysuiTestCase {
         verify(mTestCallback).onBiometricError(
                 eq(10), eq(""), eq(BiometricSourceType.FINGERPRINT));
         verify(mTestCallback, never()).onBiometricAuthenticated(
-                anyInt(), any(), anyBoolean());
+                anyInt(), any(), anyBoolean(), any(SecondFactorStatus.class));
     }
 
     @Test
@@ -911,7 +912,7 @@ public class KeyguardUpdateMonitorTest extends SysuiTestCase {
         verify(mTestCallback).onBiometricDetected(
                 eq(0), eq(BiometricSourceType.FACE), eq(false));
         verify(mTestCallback, never()).onBiometricAuthenticated(
-                anyInt(), any(), anyBoolean());
+                anyInt(), any(), anyBoolean(), any(SecondFactorStatus.class));
     }
 
     @Test
@@ -1162,6 +1163,18 @@ public class KeyguardUpdateMonitorTest extends SysuiTestCase {
         assertThat(mKeyguardUpdateMonitor.getUserCanSkipBouncer(user)).isTrue();
     }
 
+    @Test
+    public void testGetUserCanSkipBouncer_whenFingerprintAndSecondFactorEnabled_returnsFalse() {
+        int user = mSelectedUserInteractor.getSelectedUserId();
+        mKeyguardUpdateMonitor.onFingerprintAuthenticated(user, true /* isClass3Biometric */);
+        when(mLockPatternUtils.isBiometricSecondFactorEnabled(user)).thenReturn(true);
+
+        assertThat(mKeyguardUpdateMonitor.getUserHasTrust(user)).isFalse();
+        assertThat(mKeyguardUpdateMonitor.isUnlockingWithBiometricAllowedSafe(true)).isTrue();
+
+        assertThat(mKeyguardUpdateMonitor.getUserCanSkipBouncer(user)).isFalse();
+    }
+
     @Test
     public void testGetUserCanSkipBouncer_whenFingerprint_nonStrongAndDisallowed() {
         when(mStrongAuthTracker.isUnlockingWithBiometricAllowed(false /* isClass3Biometric */))
@@ -2391,6 +2404,12 @@ public class KeyguardUpdateMonitorTest extends SysuiTestCase {
         return intent;
     }
 
+    @Test
+    public void testClearFingerprintRecognized_success_clearsPendingAuthTokens() {
+        mKeyguardUpdateMonitor.clearFingerprintRecognized();
+        verify(mFingerprintManager).clearPendingAuthTokens();
+    }
+
     private void verifyFingerprintAuthenticateNeverCalled() {
         verify(mFingerprintManager, never()).authenticate(any(), any(), any(), any(), any());
         verify(mFingerprintManager, never()).authenticate(any(), any(), any(), any(), anyInt(),
-- 
2.43.0

